scripts/build.tokens.sns.ts

```
@@ -104,30 +132,33 @@ const mapOptionalToken = (response) => {
 		return acc;
 	}, {});
 
-	if (
-		isNullish(nullishToken.symbol) ||
-		isNullish(nullishToken.name) ||
-		isNullish(nullishToken.fee) ||
-		isNullish(nullishToken.decimals)
-	) {
-		return undefined;
+	const { symbol, name, fee, decimals, ...rest } = nullishToken;
+
+	if (isNullish(symbol) || isNullish(name) || isNullish(fee) || isNullish(decimals)) {
+		throw new Error(`Missing token metadata: ${JSON.stringify(response)}`);
```

Why are you changing the behavior - i.e. any particular reason to throw an error instead of returning the partial token?

One or the other, probably better to propose this in another PR given that this one has for goal to convert to typescript?

---

Ah, yes, forgot to mention in the description: this is a funny one that Typescript showed.

We were returning `undefined` in case one of the props is `nullish`. However, type-wise, it is wrong when we use the function in 

```typescript
({
	ledgerCanisterId: ledger_canister_id,
	indexCanisterId: index_canister_id,
	rootCanisterId: root_canister_id,
	metadata: {
		...mapOptionalToken(icrc1_metadata),
		alternativeName: alternativeName.trim(),
		url
	}
})
```

because the type of this object is given by schema `EnvSnsTokenSchema` once we use Typescript. And in that schema, the `metadata` provided by `mapOptionalToken` MUST be `non-nullish`. So, one solution could e the one I did, or another one is to fallback on empty metadata, i.e. `mapOptionalToken(icrc1_metadata) ?? {symbol: '', decimals: 0, ... }`

WDYT? Because as it is now, it basically ignores that they could have no metadata? Shall I change the schema, or add a new schema with partial metadata? But it is a cascade after eheh

---

👉 #4587

---

Cool! So, after that PR, I just need to make `metadata` prop optional in `EnvSnsTokenSchema`, correct?

---

Yes :)

---

Merged

---

Rebased and changed the types: have a look, i had to create a sub-function to filter because the type was not being recognized as non-nullish `metadata`

---

@AntonioVentilii it should return `undefined` instead of throwing an error no?

```
throw new Error(`Missing token metadata: ${JSON.stringify(response)}`);
```

=>

```
return undefined
```

Or am I missing something?

---

yes, indeed: i see it there, no?

---

my mistake... i forgot to push the last commit, sec

---

pushed! unfortunately I still need to deconstruct... it is giving me some type issue saying that `nullishToken` is not what we required, even after the nullish checks... strange